feat: Add file menu et menu editor option for copy url

Mara-Li opened the pull request:
I think it will be practical to have this option.

Also, do you have think about to don't show the modal if publish = false, instead of an error message?

Mara-Li committed:
feat: Add file menu et menu editor option for copy url

kometenstaub commented:
> I think it will be practical to have this option.

~~I see that you already forked this repo and added these features. Feel free to make a PR, but please let the file menu and editor options be off by default, so that they can be enabled in the settings. I don't want to clutter these menus any more than necessary.~~
~~So one toggle in the settings for the file menu and another one for the editor menu.~~


> Also, do you have think about to don't show the modal if publish = false, instead of an error message?

That is by design. If it didn't show anything, then the user might not know that the status is set to false and could think that the plugin doesn't work. (Only tags and aliases are shown by default in preview mode.)
I also see it as a kind of warning.

kometenstaub commented:
Sorry, I just realised that this was already the PR.

Please let the file menu and editor menu be optional and off by default, so that they can be enabled in the settings. I don't want to clutter these menus any more than necessary.
So one toggle in the settings for the file menu and another one for the editor menu.

kometenstaub commented on src/main.ts:
```diff
@@ -68,6 +79,47 @@ export default class CopyPublishUrlPlugin extends Plugin {
                 }
             },
         });
+        this.registerEvent(this.app.workspace.on("file-menu", (menu, file : TFile) => {
```
Please only add this if the user enabled them in the settings. (Off by default.)

kometenstaub commented on src/main.ts:
```diff
@@ -68,6 +79,47 @@ export default class CopyPublishUrlPlugin extends Plugin {
+            menu.addSeparator()
+        }));
+
+        this.registerEvent(this.app.workspace.on("editor-menu", (menu, editor, view) => {
```
Same thing here.

kometenstaub commented on src/main.ts:
```diff
@@ -27,15 +35,18 @@ export default class CopyPublishUrlPlugin extends Plugin {
-            id: 'copy-publish-url',
-            name: 'Copy URL',
-            checkCallback: (checking: boolean) => {
+            "id": 'copy-publish-url',
+            "name": 'Copy URL',
+            "checkCallback": (checking: boolean) => {
```
```suggestion
            id: 'copy-publish-url',
            name: 'Copy URL',
            checkCallback: (checking: boolean) => {
```

kometenstaub replied:
That's just a style thing.

Mara-Li commented:
I completely understand for the option, and I think it's a good idea !

Mara-Li committed:
feat(settings): Add a settings option for context menu

Mara-Li committed:
refactor: Just cleanup

Mara-Li committed:
Merge remote-tracking branch 'origin/main' into main

Mara-Li committed:
refactor: Line styling

Mara-Li committed:
refactor: Styling

Mara-Li commented:
Normally, it's all good !

kometenstaub commented on src/main.ts:
```diff
@@ -1,12 +1,20 @@
 
 const DEFAULT_SETTINGS: CopyPublishUrlSettings = {
     homeNote: '',
     publishPath: '',
```
Please add `enableContext` with a value of `false` here.

kometenstaub commented on src/main.ts:
```diff
@@ -69,6 +80,70 @@ export default class CopyPublishUrlPlugin extends Plugin {
+            menu.addSeparator()
+        }));
+
+        this.registerEvent(this.app.workspace.on("editor-menu", (menu, editor, view) => {
```
You're registering the `editor-menu` twice, but it should only appear once.

kometenstaub commented on src/settings.ts:
```diff
@@ -56,5 +56,15 @@ export default class CopyPublishUrlSettingTab extends PluginSettingTab {
+            .setDesc('Allow copy publish in context menu.')
+            .addToggle(toggle => {
+                toggle.setValue(this.plugin.settings.enableContext)
+                toggle.onChange(value =>{
```
Please make this async. `saveSettings()` calls `saveData()`, [which is async](https://github.com/obsidianmd/obsidian-api/blob/5687a75ecfdb78b6a832127cf6f9229e27e196a0/obsidian.d.ts#L2345).

kometenstaub reviewed (changes requested):
1. Please only register the two events when the setting is turned on. Currently they're getting registered even when the toggle is off, they just don't get added to the menus because of the condition check.
2. It only works if a note has `publish: true`. It shouldn't be in the menu if the note contains `publish: false`, that's correct. However, it should still appear if the `publish` flag hasn't been explicitly set to `true`.
3. Please implement two toggles, one for the editor menu, the other one for the file menu.
4. Regarding the icon, would it make more sense to use the same icon Obsidian uses for copying instead of using the paste icon? 
![image](https://user-images.githubusercontent.com/83140328/143680705-bdc2fae2-7d53-426f-8661-0ad5edd8cf95.png)
5. For the other changes, please refer to the comments I have added to the lines.

Mara-Li committed:
fix: Init enableContext with false

Mara-Li committed:
fix: fix async save + add new options

Mara-Li committed:
fix: Trying to understand file-menu / editor-menu.

Mara-Li committed:
fix: Remove editor menu ; load the event only if settings is on ; add a warning about it in option.

Mara-Li commented:
> 1. Please only register the two events when the setting is turned on. Currently they're getting registered even when the toggle is off, they just don't get added to the menus because of the condition check.

I can do that, but it force to user to always reload the plugin if they change settings. I don't think it's a good idea. It's also the same for unload, so I "invisible" the menu in case the settings is take of.
Also, I added information about reload plugin in the description option. It's up to you to keep it or remove it. 

> 2. It only works if a note has `publish: true`. It shouldn't be in the menu if the note contains `publish: false`, that's correct. However, it should still appear if the `publish` flag hasn't been explicitly set to `true`.
Done! 

>  3. Please implement two toggles, one for the editor menu, the other one for the file menu.
In fact, I was thinking that "editor menu" was for an option in context menu when a file is open, and 'file-menu' for a right click in the explorer. I realize now that is the right click **on** a file, as for copy, paste, … 
I removed this option, I don't think it's the right place to add the plugin!

> 4. Regarding the icon, would it make more sense to use the same icon Obsidian uses for copying instead of using the paste icon?
> ![image](https://user-images.githubusercontent.com/83140328/143680705-bdc2fae2-7d53-426f-8661-0ad5edd8cf95.png)

I think the `link` icon is better because Obsidian use it for "copy Obsidian URL".

Mara-Li committed:
fix: forgot to remove option

Mara-Li reviewed (commented):
Everything requested is done

kometenstaub commented:
> I can do that, but it force to user to always reload the plugin if they change settings. I don't think it's a good idea. It's also the same for unload, so I "invisible" the menu in case the settings is take of.
> Also, I added information about reload plugin in the description option. It's up to you to keep it or remove it.

I took care of that, the plugin is now automatically reloaded when the button is toggled (the settings page for this plugin is automatically reopened as well).

I also refactored the code so that both the command and the menu entry re-use the same code.

Thanks for your work on this! :)
The changes are released in 0.2.0

Mara-Li commented:
Thank you ! 🎉
